### PR TITLE
Initial support for parsing xTB output (WIP)

### PR DESCRIPTION
**Pull request by Shualdon**
Hi,

As discussed in #789, I have added initial support for reading xTB output files. 
Currently it only supports GFN2 calculations using the latest xTB version (6.4.1).  It can probably read older versions, but there are some changes in the text it won't be able to parse, for now.
I added the "atomprop" and "bondprop" attributes ([#789](https://github.com/cclib/cclib/issues/789#issuecomment-1018114031)), and tried to use as much of the existing attributes as possible.
I tried as much as I can to follow the same style as the other parsers.
I also added some example output files.
This is a WIP, so I would be happy to get feedback. 

Omri.

### CODE REVIEW

**Comment by ghutchis** [comment]
I guess the next step would be to add some test coverage? Is there a list of the basic systems to run to add to the data files repo?

File: cclib/parser/xtbparser.py [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+        # 18
+        # xtb: 6.3.0 (007a174)
+        # C        -1.06986995496938    0.00558244661493   -0.31957205560407
+        # O        -0.17255304749241   -0.97627397755570   -0.37573083147008
+        # ...
+        # H         1.66147402211077    1.54957926818092    0.61042104923243
+        # H         5.25469642314449   -2.38603632440806    0.38612072928186
+        # H         3.38728187535945   -4.19429919486669   -0.44692022446750
+        #
+        # For sdf\mol input file - gives sdf\mol output coords
```
[the remark this answers is not in the document]
Comment (ghutchis) [reply cclib/parser/xtbparser.py:102]: I guess the parser has to guess based on format after the `final structure:` line?

[the remark this answers is not in the document]
Comment (awvwgk) [reply cclib/parser/xtbparser.py:102]: Guessing formats can be difficult, therefore I stopped guessing the user input format but only rely on the name of the input file instead in `xtb`. If it is helpful `xtb` could write the detected input format as well.

File: cclib/parser/xtbparser.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+        # ........................................................................
+        #    1    -37.9590419 -0.379590E+02  0.116E-04    2.97       0.0  T
+        #    2    -37.9590419 -0.674305E-11  0.982E-05    2.97     240.0  T
+        #    3    -37.9590419 -0.129340E-09  0.425E-05    2.97     554.7  T
+        #      SCC iter.                  ...        0 min,  0.004 sec
+        #      gradient                   ...        0 min,  0.002 sec
+        #  * total energy  :   -37.3373741 Eh     change       -0.1278963E-08 Eh
+        #    gradient norm :     0.0004128 Eh/α   predicted     0.0000000E+00 (-100.00%)
+        #    displ. norm   :     0.0058476 α      lambda       -0.2852929E-06
+        #    maximum displ.:     0.0054052 α      in ANC's #1, #7, #3, ...
```
Comment (awvwgk) [inline cclib/parser/xtbparser.py:69]: This output is dependent on the optimizer used, `xtb` currently supports two variants of ANCopt and the RF based one is exchanged for the LBFGS based at 500 atoms. Also, the user can configure this in the input file always use the RF or LBFGS based version.

File: cclib/parser/xtbparser.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+        #
+        #   program call               : <command>
+        #   hostname                   : <hostname>
+        #   coordinate file            : <filename>.<<filetype>
+        #   omp threads                :                    12
+        #   number of atoms            :                    18
+        #   number of electrons        :                    66
+        #   charge                     :                     0
+        #   spin                       :                   0.0
+        #   first test random number   :      0.87181443679343
```
Comment (awvwgk) [inline cclib/parser/xtbparser.py:41]: This output will be gone in the next `xtb` release.

File: cclib/parser/xtbparser.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+                self.set_attribute('atomnos', atomnos)
+                self.set_attribute("atomcoords", atomcoords)
+            
+            else:
+                pass
+
+
+        # Get Molecular Orbitals energies and HOMO index
+        # xTB trunctaes the MO list so we need to take care of that.
+        # Unkown energies will be given NaN as a value
```
Comment (awvwgk) [inline cclib/parser/xtbparser.py:166]: Try running this part of the parser with higher electronic temperature `--etemp 5000` or so.

File: cclib/parser/xtbparser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shualdon) [reply cclib/parser/xtbparser.py:102]: This is how I did it here - guessing the format from the input file (as the output prints it). I only did `xyz `and `mol\sdf `formats so far, as those are the formats I mostly use, but I plan to add more in the future.
Having `xtb ` print the format can be useful, but I guess it's going to work similarly to how it is done now.

[the remark this answers is not in the document]
Comment (awvwgk) [reply cclib/parser/xtbparser.py:102]: An easy way to match the detection logic in `xtb` is to collect the coordinate file from the initial printout and reproduce the suffix / basename selection rules for the file format from the IO library:

https://github.com/grimme-lab/mctc-lib/blob/4e88c5239344dad805c91ae021694891ff0dd372/src/mctc/io/filetype.f90#L71-L128

[the remark this answers is not in the document]
Comment (Shualdon) [reply cclib/parser/xtbparser.py:102]: That's basically how I did it. I just need to add support to more formats.

[the remark this answers is not in the document]
Comment (Shualdon) [reply cclib/parser/xtbparser.py:102]: That can be done. I will work on it.

**Comment by ghutchis** [comment]
FYI - https://github.com/nomad-coe/nomad-parser-xtb

**Comment by berquist** [comment]
Hi Omri, is there anything we can do to move this forward?  It looks like a lot of work is already done and it shouldn't go to waste.

**Comment by Shualdon** [comment]
Hi, sorry it's been long since I updated. My research is keeping me busy. I would continue to work on this and will have a new PR as soon as I can.

**Comment by berquist** [comment]
I will hand-check the output, since there are no tests, and if all is good I'll merge what's here for 1.8.

**Comment by berquist** [comment]
Only the LMO examples are parsing right now, giving the following attributes:
```
  atomprop
  bondprop
  charge
  homos
  metadata
  moenergies
  mult
```

**Review by shivupa — APPROVED** [review]
This seems good for now. Future efforts should include

- better description of atomprop/bondprop in ccdata
- abstract out geometry parsing
- based on @awvwgk's review (thank you btw!) this might fail geom opts with different optimizers.
We can open an issue for future xtb parsing development

**Comment by Shualdon** [comment]
Hi everyone, thanks for the feedback and work on getting this PR merged! I'll work on improving it and submit another PR when it's ready.

**Comment by shivupa** [comment]
Hi Omri! I think the plan will be to merge this (as long as there are no objections). The list I included is for future PRs.

**Comment by berquist** [comment]
I will fix the merge conflict and try to get everything here parsing then merge for 1.8.

**Comment by Andrew-S-Rosen** [comment]
@berquist and friends: What's the current status of this PR? Looks like all tests pass but just merge conflicts have to be resolved? 

If this is in a fairly respectable state as-is, I'm happy to take a stab at expanding on it (+ adding tests) via a side-project I'm hoping to start soon. That might be best suited as a set of follow-up PRs though because I haven't quite decided yet when I'm going to start this effort.

**Comment by shivupa** [comment]
I think having a interested user indicates to me that this should be merged asap + add improvements later.

I am ok with that to move things along on your end @Andrew-S-Rosen 

Sorry I realize this exactly what I said in this thread months ago, but this slipped through the cracks with version 2 development going on.. (also coming soon™️)

**Comment by Andrew-S-Rosen** [comment]
@shivupa: No need to apologize. We've all been there! That's just how it goes.

I'd certainly be more inclined to dig in a bit if there's interest in getting this merged 😄 I'm going to take a stab at making an ASE-based xTB calculator using cclib for the output parsing and jinja for the input setup, mostly as an excuse to learn jinja and find areas for improvement in the cclib parser. Will probably give that a spin starting sometime next week, if all goes according to plan.

**Comment by Andrew-S-Rosen** [comment]
@shivupa: Alright, I'm committed to this know! 😉 

I installed the PR and ran the most recent version of xTB, which produced an `xtb.out` file that I read with `ccread`.

First, the good news: no errors were raised! 🥳 

Next the not great news: the resulting `ccData` object is missing many goodies! For instance, I don't see any `scfenergies` or `grads`.

Reproduction:
```python
from cclib.io import ccread
cclib_obj = ccread("xtb.out")
print(cclib_obj.getattributes())
```

[xtb.out.txt](https://github.com/cclib/cclib/files/13173813/xtb.out.txt)

Returns the following:
```python
{'atomprop': {'convcn': [0.908, 0.908],
  'q': [0.0, -0.0],
  'c6aa': [3.062, 3.062],
  'alpha': [2.737, 2.737]},
 'bondprop': {'wbo': [[1.0, [1, 1.0]], [1.0, [0, 1.0]]]},
 'homos': array([0], dtype=int32),
 'metadata': {'package': 'XTB',
  'methods': [],
  'success': True,
  'legacy_package_version': '6.6.1',
  'coord_type': 'xyz'},
 'moenergies': [array([-12.8334,   5.2022])]}
 ```
 
So, there's certainly some work to be done! But I'm hopeful it won't be too painful. Perhaps an update to the xTB output file caused a parsing mismatch. I haven't looked at this PR's code yet. I'll keep you posted.

**Comment by berquist** [comment]
This is superseded by #1296.  Andrew has made sure to keep Omri's commits and we will not squash the full branch.

**Pull request closed without merging**